### DISCUSSION TITLE
Adaptive H3 grid

### DISCUSSION THREAD

**Discussion by ysig** [uber/h3#737] [Q&A] [answered]
I'm trying to use h3 in the following way:

I have a set of data points with lat,long coordinates. Although I can connect any of them with a cell number I want to organize them in a way so that for each cell the number of points is balanced - as with an adaptive grid:
![image](https://user-images.githubusercontent.com/28439529/206928137-585ee8d6-624d-4aca-b0f9-8cb43675f7e9.png)

Is this supported or possible with H3?

**Comment by nrabinowitz** [comment, marked as answer]
This is definitely possible with H3, though H3 doesn't support it directly. H3 has a hierarchical index structure that can be iteratively subdivided to get an adaptive grid like the one in your image. Here's an example notebook to demonstrate: https://observablehq.com/@nrabinowitz/h3-adaptive-grid

There are various ways you can construct the grid. The one I use here assigns each point to a res 15 cell, then walks up the hierarchy until it finds a parent cell with other points in it. If the number of points in that parent cell is now beyond the max point threshold, it subdivides the parent and redistributes the points.

Note that, because H3 cells do not perfectly contain their children, there are a few discrepancies when using H3 instead of S2:
- Points may be allocated to coarse cells that do not strictly contain them
- There are visual overlaps between cells in a multi-resolution grid

Depending on your use case, these issues may or may not be blockers.

**Comment by ysig** [comment]
That's amazing! Thank you so much!
